Move & Copy Assignment for BinarySpaceTree, Octree and CosineTree

sriramsk1999 opened the pull request:
This is for resolving issue #1957 , or atleast a part of it. 

I've started out with BinarySpaceTrees referring to the implementation in #1934 for RectangleTrees. I've mostly replicated the code written by @rcurtin but I want to know if I'm on the right track. 

In short:
 Added move and copy assignment operators for BinarySpaceTree
 Added a test case for both operators 
 Included name in the contributors list :)

walragatver commented:
Looks good. You can move ahead. :+1:

rcurtin commented on src/mlpack/core/tree/binary_space_tree/binary_space_tree_impl.hpp:
```diff
@@ -379,6 +379,102 @@ BinarySpaceTree(
+  parentDistance = other.ParentDistance();
+  furthestDescendantDistance = other.FurthestDescendantDistance();
+  minimumBoundDistance = other.MinimumBoundDistance();
+  dataset = other.dataset;
```
Don't forget to set the values of `other` to their default values. :)

sriramsk1999 commented:
I have set parameters of `other` to their defaults. In addition, I noticed that the Move assignment for Rectangle tree didn't change reset `other` parameters so I added that as well. 

And in `binary_space_tree_impl.cpp`, along with resetting `other` I included `other.parent=NULL` as well. Would like these changes reviewed

rcurtin commented:
@sriramsk1999 I took a look at the code and it looks good to me (thank you for adding the tests too).  Do you want to also handle the cosine tree and the octree also in this PR?

sriramsk1999 commented:
I'll open another PR for those two trees, and CoverTree and SpillTree too if they aren't done already. Readying this for review and renaming the PR to reflect the commits :)

walragatver commented on src/mlpack/core/tree/binary_space_tree/binary_space_tree_impl.hpp:
```diff
@@ -406,6 +512,7 @@ BinarySpaceTree(BinarySpaceTree&& other) :
   // tree's contents, so it doesn't delete anything when it is destructed.
   other.left = NULL;
   other.right = NULL;
+  other.parent = NULL;
```
Nice catch.

walragatver commented on src/mlpack/tests/knn_test.cpp:
```diff
@@ -1382,6 +1410,33 @@ BOOST_AUTO_TEST_CASE(MoveConstructorRTreeTest)
+  knn->Search(3, neighbors, distances);
+
+  // Use move constructor.
+  NeighborSearchType knn2(std::move(*knn));
```
It looks like we are only verifying `move constructor`. However, we should also check `move assignment`. To do that below is the code (Add it after knn2.search())
```
NeighborSearchType knn3 = std::move(*knn);
knn3.search(neighbors3, distances3);
``` 

It looks like this also got missed in #1934. So it would be great if you modify that test case as well.
For more reference:
https://en.cppreference.com/w/cpp/language/move_assignment

walragatver commented on src/mlpack/tests/knn_test.cpp:
```diff
@@ -1405,13 +1405,17 @@ BOOST_AUTO_TEST_CASE(MoveConstructorRTreeTest)
+  knn3.Search(3, neighbors3, distances3);
+
   CheckMatrices(neighbors, neighbors2);
   CheckMatrices(distances, distances2);
```
Looks like you have added the code but you failed to check the matrices.

sriramsk1999 replied:
Oh wow, can't believe I missed that. Will add it in :)

sriramsk1999 commented:
I have added the move/copy constructors/assignment for CosineTree but this is a work in progress. Would appreciate a little help :) In particular:

- Why is the structure of the `cosine_tree` different compared to other folders. For other trees I have been adding the implementation in the `tree_name_impl.hpp` whereas there is only a `cosine_tree.cpp` in this folder.

- How do I traverse through the tree? Looked at some literature, was unable to find anything useful. This is to copy/move the children of the tree

- Another intriguing error is the commented out line about moving the dataset parameter. My IDE highlights the assignment and I do not know why.

Help would be appreciated :)

walragatver commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,96 @@ CosineTree::CosineTree(const arma::mat& dataset,
+  if (this == &other)
+    return *this;
+
+  //dataset = (other.parent == NULL) ? other.dataset : NULL;
```
Dataset is a reference over here not a pointer. :)

sriramsk1999 replied:
I'm not sure I follow. I saw that the datatype is `const arma::mat` instead of `MatType*` like it is for the other trees. I read up on C++ references but I can't figure out why I can't assign the dataset reference. 
How do I go about assigning a reference?

walragatver replied:
https://www.geeksforgeeks.org/when-do-we-use-initializer-list-in-c/
This will answer your doubt. 2nd point.

walragatver replied:
Here's the correct code for making a perfect copy (deep copy)
use the initializer list to remove the error. 
dataset (other.getdataset())

Then in the constructor body 
```
arma::mat a = other.getDataset();

this->dataset = a;
```

sriramsk1999 replied:
It's still not working. I've changed `dataset((other.parent == NULL) ? other.dataset : NULL)` to `dataset((other.parent == NULL) ? other.GetDataset() : NULL)` in the Copy Constructor's initializer list.

> Then in the constructor body

Did you mean to say in the body of Copy Assignment? Nevertheless, I tried adding the two lines you've specified in both the body of the Constructor and the Assignment but the error didn't go away.

walragatver replied:
> It's still not working. I've changed `dataset((other.parent == NULL) ? other.dataset : NULL)` to `dataset((other.parent == NULL) ? other.GetDataset() : NULL)` in the Copy Constructor's initializer list.

What's the error the initializer list is throwing?

sriramsk1999 replied:
The initializer list doesn't show any error, with or without GetDataset(). It's the assignment in Copy and Move assignment which shows the error.

The error is `no operator "=" matches these operands -- operand types are: const arma::mat = const arma::mat`

walragatver replied:
Okay fine. That error is because of using a `const` object reference as it will need a `const` function. So I would say it will be good if we make `dataset` member as `non-const reference` inside `cosine_tree.hpp`. 
However, the author would have kept it `const` because of some reason. So let us know if it creates an error again after making the said change.

sriramsk1999 commented:
I have modified `dataset` to be a non-const reference in both `cosine_tree.hpp` and `cosine_tree.cpp`. Also included is the code to copy children of the tree (from `cosine_tree_test.cpp`) and setting the parameters of `other` to their defaults in Move Assignment/Constructor. 

 -  Using `data.col()` in the Copy Assignment/Constructor functions shows up as an error on the IDE stating `a namespace name is not allowed`.
 - In the initializer list of Copy Constructor, `dataset((other.parent == NULL) ? other.dataset : NULL)` shows an error `initial value of reference to non-const must be an lvalue`. I assumed this was because of NULL, however `dataset = (other.parent == NULL) ? other.dataset : NULL;` in Copy Assignment does not show the same error.
 - Finally, in Move Assignment/Constructor, `other.dataset` cannot be set to NULL as it is a reference. How do I set a reference to its default value?

walragatver commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,209 @@ CosineTree::CosineTree(const arma::mat& dataset,
 
+//! Copy the given tree.
+CosineTree::CosineTree(const CosineTree& other) :
+  //dataset((other.parent == NULL) ? other.dataset : NULL)
```
For this. You just do `dataset(other.dataset)` temporarily in the `initializer list` (to remove the initialization error) and then do `deep copy` as mentioned in my previous comments.

walragatver replied:
https://github.com/mlpack/mlpack/pull/2023#discussion_r331147450

sriramsk1999 commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,209 @@ CosineTree::CosineTree(const arma::mat& dataset,
+      nodeStack.push_back(currentRight);
+
+      // Obtain the split point of the popped node.
+      arma::vec splitPoint = data.col(currentNode->SplitPointIndex());
```
Figured this out. It was because I was referring to `cosine_tree_test.cpp` which has declared `data` instead of `dataset`. Should I change the variable name in that file as well for consistency?

sriramsk1999 commented:
Instead of doing `dataset(other.dataset)` I used `dataset(other.parent->GetDataset())`. I assume this workaround will do a deep copy instead of a shallow copy.

rcurtin commented on src/mlpack/core/tree/octree/octree_impl.hpp:
```diff
@@ -363,6 +363,37 @@ Octree<MetricType, StatisticType, MatType>::Octree(const Octree& other) :
+  // parent links are set right.
+  for (size_t i = 0; i < other.NumChildren(); ++i)
+  {
+    children.push_back(new Octree(other.Child(i)));
```
`children` hasn't been cleared before these calls, and also when `children` is cleared we need to be sure that no memory leaks.

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,210 @@ CosineTree::CosineTree(const arma::mat& dataset,
+  // Now we are a clone of the other tree.  But we must also clear the other
+  // tree's contents, so it doesn't delete anything when it is destructed.
+  arma::mat a;
+  other.dataset = a;
```
I'd suggest `other.dataset = arma::mat();`.

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,210 @@ CosineTree::CosineTree(const arma::mat& dataset,
+  other.parent = NULL;
+  other.left = NULL;
+  other.right = NULL;
+  other.splitPointIndex = ColumnSampleLS();
```
Shouldn't we set this to `0`?

rcurtin commented on src/mlpack/core/tree/octree/octree_impl.hpp:
```diff
@@ -389,6 +420,42 @@ Octree<MetricType, StatisticType, MatType>::Octree(Octree&& other) :
+  if (this == &other)
+    return *this;
+
+  children = std::move(other.children);
```
We need to clear the memory for the existing `children` safely before this.

rcurtin commented on src/mlpack/core/tree/rectangle_tree/rectangle_tree_impl.hpp:
```diff
@@ -337,12 +351,28 @@ operator=(RectangleTree&& other)
   auxiliaryInfo = std::move(other.auxiliaryInfo);
 
+  // Now we are a clone of the other tree.  But we must also clear the other
+  // tree's contents, so it doesn't delete anything when it is destructed.
```
Nice catch with this part. :+1:

sriramsk1999 commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -17,7 +17,7 @@
 namespace tree {
 
-CosineTree::CosineTree(const arma::mat& dataset) :
+CosineTree::CosineTree(arma::mat& dataset) :
```
After [this](https://github.com/mlpack/mlpack/pull/2023#discussion_r330858718) discussion with @walragatver I changed `dataset` to a non-const member.

rcurtin replied:
Got it, thanks for the explanation.  I see that in order to make these copy constructors and operators work, we need to be able to "re-seat" the reference.  The way that you've done that is by using `std::move()`.  That works, but the only thing is, it causes a user to expect their data to be modified (since the parameter is not `const`).  Also, it makes serialization much harder when the object holds a reference.  (I can see that serialization isn't implemented for the cosine tree, and I bet this is exactly why.)

So, I wanted to ask if you'd be willing to improve the `CosineTree` class by changing the member `dataset` from a reference to a pointer (i.e. `arma::mat* dataset`).  Then it would be like `BinarySpaceTree` and the other tree classes, and it would be much easier to implement the copy and move constructors and operators.  However, there would be a little additional work to make sure that the pointer was being properly handled, and we'd need to add a destructor.

We could use the same ownership rules as `BinarySpaceTree`: if `parent == NULL`, then we are responsible for deleting `dataset`.

Let me know what you think. :)

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,197 @@ CosineTree::CosineTree(const arma::mat& dataset,
+}
+
+//! Move Assignment.
+CosineTree&CosineTree::operator=(CosineTree&& other)
```
```suggestion
CosineTree& CosineTree::operator=(CosineTree&& other)
```

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,197 @@ CosineTree::CosineTree(const arma::mat& dataset,
+
+//! Move the given tree.
+CosineTree::CosineTree(CosineTree&& other) :
+  dataset(other.dataset),
```
```suggestion
    dataset(other.dataset),
```

Can you make these lines indented doubly?  They are wrapped from the previous line, so that's what the style guide suggests. :+1:

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,197 @@ CosineTree::CosineTree(const arma::mat& dataset,
+}
+
+//! Copy Assignment.
+CosineTree&CosineTree::operator=(const CosineTree& other)
```
```suggestion
CosineTree& CosineTree::operator=(const CosineTree& other)
```

rcurtin commented on src/mlpack/core/tree/rectangle_tree/rectangle_tree_impl.hpp:
```diff
@@ -256,6 +263,9 @@ RectangleTree(RectangleTree&& other) :
 }
 
+/**
+ * Copy Assignment
```
```suggestion
 * Copy assignment operator: copy the given other tree.
```

Sorry to be a little pedantic, but I wanted to add a bit about what this method does for clear documentation's sake.

rcurtin commented on src/mlpack/core/tree/rectangle_tree/rectangle_tree_impl.hpp:
```diff
@@ -304,6 +316,9 @@ operator=(const RectangleTree& other)
 }
 
+/**
+ * Move Assignment
```
```suggestion
 * Move assignment operator: take ownership of the given tree.
```

Same here. :+1:

sriramsk1999 commented:
Hey @rcurtin , no worries about the delay 🙂 I'll get round to addressing your comments as soon as I can, but I'm swamped with work right now (end of the semester 😞 ) so I apologize in advance for the delay.

rcurtin commented:
Hey @sriramsk1999, no worries---hope the end of the semester is going well. :)  If there are any comments I can clarify or help with, just let me know. :+1:

sriramsk1999 commented:
Hey @rcurtin , I think I covered most of what you pointed out. By my reckoning, only a few things left for the CosineTree:

- Test for Move/Copy Operators
-  Destructor
 - Converting dataset to a pointer

I'm focusing on the last part first and I'm having a bit of trouble. I get numerous errors because of calls like `dataset.n_cols` and I'm not sure how to fix it. In other trees, the `dataset` member is a `MatType*` and not `arma::mat*` and I was wondering if this had anything to do with it.

rcurtin commented:
Hey @sriramsk1999, you can't use the `.` operator with pointers.  Since `dataset` is `MatType*` and not `MatType&`, you'll need to do `dataset->n_cols`.  Hope that helps. :+1:

Looking forward to seeing what you come up with. :+1:

sriramsk1999 commented:
> Hey @sriramsk1999, you can't use the `.` operator with pointers. Since `dataset` is `MatType*` and not `MatType&`, you'll need to do `dataset->n_cols`. Hope that helps. +1

Oh I can't believe I missed that, that's embarassing 😅

 - Made dataset a pointer member of CosineTree.

 - In QUIC_SVD and PCA I made dataset a non-const reference as it would be referenced when calling CosineTree constructors.

Would like these changes reviewed 🙂

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -17,12 +17,12 @@
 namespace tree {
 
-CosineTree::CosineTree(const arma::mat& dataset) :
+CosineTree::CosineTree(arma::mat* dataset) :
```
Let's not change the API here.  Holding a pointer to the dataset internally is just fine, but take a `const arma::mat&` as the input please. :+1:  (You may need to hold a const pointer internally, then.)

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,196 @@ CosineTree::CosineTree(const arma::mat& dataset,
 
+//! Copy the given tree.
+CosineTree::CosineTree(const CosineTree& other) :
+    dataset(other.parent->GetDataset()),
```
Why `.parent` here?

sriramsk1999 replied:
Changed it to `other.dataset`.

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,196 @@ CosineTree::CosineTree(const arma::mat& dataset,
+    delta(other.delta),
+    parent(other.Parent()),
+    left(other.Left()),
+    right(other.Right()),
```
Since you're modifying these later, you could just set them to `NULL` here.  I think that would be more clear.

sriramsk1999 replied:
Do you mean `delta` as well or just `left`,`right` and `parent`? Because `delta` isn't being modified from what I can see.

rcurtin replied:
Oh, just `left`, `right`, and `parent`.  It looks good now, thanks. :+1:

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,196 @@ CosineTree::CosineTree(const arma::mat& dataset,
+  other.splitPointIndex = 0;
+  other.numColumns = dataset->n_cols;
+  other.l2Error = -1;
+  other.frobNormSquared = arma::accu(l2NormsSquared);
```
It doesn't make sense to me why you are setting parts of `other` to have values relevant to this tree.

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,196 @@ CosineTree::CosineTree(const arma::mat& dataset,
+  other.splitPointIndex = ColumnSampleLS();
+  other.numColumns = dataset->n_cols;
+  other.l2Error = -1;
+  other.frobNormSquared = arma::accu(l2NormsSquared);
```
Here too, it's not clear why these should be the default values.

sriramsk1999 replied:
That's strange... I distinctly remember referring to another part of the codebase to set these values. But when I searched I can't find anything.  

For now I've set them to zero which I assume to be default values. (Other than `l2Error` which is initialized to -1 in another constructor.)

rcurtin replied:
Got it, no worries. :+1:

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.hpp:
```diff
@@ -169,7 +198,7 @@ class CosineTree
 
   //! Get pointer to the dataset matrix.
-  const arma::mat& GetDataset() const { return dataset; }
+  arma::mat* GetDataset() { return dataset; }
```
Let's leave this as a reference also so the external API for the class does not change.

rcurtin commented on src/mlpack/core/tree/rectangle_tree/rectangle_tree_impl.hpp:
```diff
@@ -183,6 +183,7 @@ RectangleTree(
     maxLeafSize(other.MaxLeafSize()),
     minLeafSize(other.MinLeafSize()),
     bound(other.bound),
+    stat(other.stat),
```
Nice catch!  Thank you. :+1:

rcurtin commented on src/mlpack/methods/pca/decomposition_policies/quic_svd_method.hpp:
```diff
@@ -50,8 +50,8 @@ class QUICSVDPolicy
-  void Apply(const arma::mat& data,
-             const arma::mat& centeredData,
+  void Apply(arma::mat& data,
+             arma::mat& centeredData,
```
Once we change the `CosineTree` to take a reference again, this change should not be needed.

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,209 @@ CosineTree::CosineTree(const arma::mat& dataset,
+  if (this == &other)
+    return *this;
+
+  dataset = (other.parent == NULL) ? other.dataset : NULL;
```
@sriramsk1999 this one does still need to be fixed. :+1:  It turns out, it has some other implications too: `CosineTree` will now need to track whether it is responsible for deleting the `dataset` pointer when it is deleted.  I didn't realize that earlier, sorry about that.  It will take just a little bit of refactoring but shouldn't be too hard I think. :)

rcurtin replied:
This comment also has not been addressed.

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,6 +150,196 @@ CosineTree::CosineTree(const arma::mat& dataset,
 
+//! Copy the given tree.
+CosineTree::CosineTree(const CosineTree& other) :
+    dataset(other.dataset),
```
The same copying issue as in @walragatver's comment will apply here too. :+1:

sriramsk1999 replied:
I'm not entirely sure if these are the fixes you meant for but they seem correct 🙂

sriramsk1999 replied:
Hmm they definitely aren't correct. Could you go into a bit more detail about how I should be handling the deletion of the `dataset` pointer?

rcurtin replied:
Well, so there are a couple options.  The one I suggested is to hold a `bool` inside the class indicating whether or not the `CosineTree` owns the `dataset` pointer (and thus is responsible for calling `delete` on it in the destructor).

The other approach you could take is to adapt the `CosineTree` to be more like, e.g., the `BinarySpaceTree`, where the root node *always* owns the dataset and must free it in the destructor.  That will be slightly more work though since you will have to adapt the code that uses the `CosineTree`.

sriramsk1999 replied:
> Well, so there are a couple options. The one I suggested is to hold a `bool` inside the class indicating whether or not the `CosineTree` owns the `dataset` pointer (and thus is responsible for calling `delete` on it in the destructor).

Ah I see. Like the `localDataset` variable in one of the other trees. Will do.

rcurtin replied:
This still has a problem.  Suppose that I do the following code:

```
arma::mat dataset(100, 100, arma::fill::randu);
CosineTree* ct = new CosineTree(dataset);
CosineTree copy(*ct); // make a copy of `ct`
delete ct; // deletes `ct.dataset`
// now copy.dataset is invalid!
```

So the copy constructor should copy everything---including the dataset.  Therefore this particular line will need to be updated.

sriramsk1999 replied:
https://pastebin.com/rP7kZeQc

I ran the above sample and it worked fine. What do you think? I understand what you said but it works regardless

rcurtin replied:
Deep copies aren't being made when the pointer is copied. Try running that code with `valgrind` and you should see that there is an invalid memory access. :+1:

sriramsk1999 replied:
Could you take a look now @rcurtin ?

rcurtin replied:
Ah, I didn't see the edit when I responded this morning:

> Edit2: So I ran a few more tests and I found out a few things. A deep copy isn't taking place (which should have been obvious actually 😅). copy.dataset tree points to the same location as ct->dataset. Calling delete ct sets ct->dataset to NULL but does not delete the dataset variable itself, allowing the copied tree to access the dataset.
>
> I assume this is because the dataset variable is allocated on the stack and thus not deleted. Is this okay or should I be deleting it somehow?

Ah I realized there is a problem in my example code.  Here we actually need to do this:

```
arma::mat* dataset = new arma::mat(100, 100, arma::fill::randu);
CosineTree* ct = new CosineTree(dataset);
CosineTree copy(*ct); // make a copy of `ct`
delete dataset;
// now copy.dataset is invalid!
```

Basically the issue with my previous code snippet was that `ct` does not "own" `ct->dataset` so it will not delete it.  If `ct` *does* own `ct->dataset`, or if `dataset` is owned by something else and then deleted, then `copy.dataset` will be invalid.  So when we use the copy constructor, we need to make sure that the `CosineTree` we're creating will not have this problem, and that includes copying the dataset.  Let me know if I can clarify that.

sriramsk1999 replied:
Oh okay I get it, that makes sense

sriramsk1999 commented:
@rcurtin for some reason, bin/mlpack_test is not working. I've been trying to debug it and I've had no luck. I asked twice on IRC and didn't get any reply. I've just reverted my changes for now, trying to get mlpack_test fixed before proceeding further.

So basically what happens is, after building when I run `bin/mlpack_test` I get:

`bin/mlpack_test: symbol lookup error: bin/mlpack_test: undefined symbol: _ZN6mlpack4tree10CosineTreeC1ERKN4arma3MatIdEE`

At first, I assumed it was because of my changes so I reverted those and tried again. No luck. After that I even made a fresh clone of the public repo and tried running the tests. Still the same error. Do you have any idea what might be happening?

rcurtin commented:
Hi @sriramsk1999, what have you tried to debug the issue?  The output is telling you that the symbol `mlpack::tree::CosineTree(const arma::mat&)` is undefined.  That typically means that the implementation of that method is not available in any of the libraries that the program (which here is `bin/mlpack_test`) is linked to.  If you go and find some resources about how linking works in C++, I think that maybe the issue here will make a little more sense and you can work out the right way to solve it. :+1:

sriramsk1999 commented:
@rcurtin I finally figured it out.  

I ran `nm -D lib/libmlpack.so.3 | grep "_ZN6mlpack4tree10CosineTreeC1ERKN4arma3MatIdEE"`  
and I didn't understand why bin/mlpack_test was throwing an error when the symbols were actually defined. 
 
After running `ldd -r bin/mlpack_test` I noticed that libmlpack.so.3 was being called on from `/usr/local/liib` and not the lib directory inside mlpack. Sure enough, the cosine tree symbols were not defined in `/usr/local/lib/libmlpack.so.3`

After that I just copied over the new shared library from inside mlpack to the one in /usr/local/lib  😊

I have a couple of questions I was hoping you could clear up for me:

1. Did this happen because I changed the api?
2. When I rebuilt mlpack, shouldn't the library stored in /usr/local/lib be overwritten with the new one?

Thanks for the help!

sriramsk1999 commented:
@rcurtin I think `localDataset` is fixed now, and I started work on the tests.

I wrote a pretty basic test and it is failing. Is there anything wrong with the test itself or is it the Copy operators?

sriramsk1999 commented:
I'm itching to have this merged :) Any final comments @rcurtin ?

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,8 +153,218 @@ CosineTree::CosineTree(const arma::mat& dataset,
+{
+  // Performing a deep copy of the dataset.
+  arma::mat cpy = other.GetDataset();
+  dataset = &cpy;
```
This is actually even more unsafe---`cpy` is a local variable and thus will go out of scope at the time this constructor ends, leaving `dataset` a pointer to deallocated memory.  I think maybe it would be helpful to spend some time reading about heap and stack allocations and memory management in C++---if you search for some nice resource, I am sure there are places that can explain it better than I can.

In our case, what we want to do is create a new stack-allocated matrix that has copied all the contents of `other.dataset` to use as `dataset`, so, in the initialization list, we can use `dataset(new arma::mat(*other.dataset))`, and then we have to be sure that `localDataset` is set correctly.

sriramsk1999 replied:
Got it, will do. The reason I took that approach was because @walragatver specified that in one his earlier comments 
 (https://github.com/mlpack/mlpack/pull/2023#discussion_r331147450)

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,8 +153,213 @@ CosineTree::CosineTree(const arma::mat& dataset,
+    numColumns(other.NumColumns()),
+    l2Error(other.L2Error()),
+    frobNormSquared(other.FrobNormSquared()),
+    localDataset(other.localDataset && other.parent == NULL)
```
This isn't correct. `localDataset` indicates whether this `CosineTree` is responsible for deleting `dataset`. What you've written here does not match the initialization of `dataset` above.

sriramsk1999 replied:
I'm not sure I follow. Are you saying that it should be initialized to `other.parent==NULL`? Put another way, will the copied tree be responsible for deleting its dataset even if the tree it copies from is not responsible for deleting its dataset? 
Sorry about the repeated questions I just want to make sure I fully understand what's going on. :)

rcurtin replied:
If the `CosineTree` has allocated the memory then it is responsible for deleting it. So, in this case, the copy constructor and copy operators are allocating new memory for the dataset and therefore that `CosineTree` object is responsible for deleting it.

rcurtin replied:
Ok, now we're getting closer with 7a8814f.  But now we need a bit of specialized behavior.  In a copy constructor, we have two situations: (a) the node is the root node, and (b) the node is not the root node.  If it's situation (a), then we need to copy the dataset and set `localDataset` to `true` so that the dataset is freed when the `CosineTree` is deallocated.  The code appears to do that now (I need to double-check but I'm pretty sure).

If it's situation (b) though, we need to *avoid* copying the dataset, as we are only holding a copy of the dataset in the root of the tree.  i.e., for a child node, `this->dataset == this->parent.dataset`, but the child node is not responsible for deleting the dataset, so `localDataset` should be `false`.  The code currently appears to set `localDataset` correctly, but `this->dataset` will still make a copy in this case.

In other trees, the way that this is done is that we manually set `left->dataset` and `right->dataset` after calling the copy constructor / constructor to create children.  Take a look at `BinarySpaceTree` and you'll see what I mean.

Thanks for persisting with this---I hope that my help has made sense and illuminated some details about pointers and allocations, and if not, don't hesitate to ask me to clarify. :+1:

sriramsk1999 replied:
(a) Yes, I'm fairly sure that it ensures required behaviour there.

(b) I did think of the same issue but I ran this [sample](https://pastebin.com/YgtEX2Z8) and found that all the nodes of the tree point to the same memory location. I didn't quite understand why but seeing as it did what was needed I left it. 

So what am I doing wrong here?

sriramsk1999 replied:
Oh I figured it out. I am in fact manually setting the dataset pointer just like in the `BinarySpaceTree`. However this assignment comes after allocating memory leading to a massive memory leak which I just found through `valgrind`. 

I'm curious, why didn't the Jenkins Memory Check job fail? Shouldn't it catch something like this?

rcurtin commented on src/mlpack/tests/cosine_tree_test.cpp:
```diff
@@ -222,4 +222,215 @@ BOOST_AUTO_TEST_CASE(CosineTreeModifiedGramSchmidt)
+
+  // Copy constructor and operator.
+  CosineTree ctree2(ctree1);
+  CosineTree ctree3 = ctree1;
```
The reason that the memory checks didn't fail is because we aren't deallocating `ctree1` here or making its dataset invalid. To make this test better, you could use `new` to allocate both `ctree1` and `data`, and then `delete` both of those directly after the copies above are done. Can you make that change? :)

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,8 +153,214 @@ CosineTree::CosineTree(const arma::mat& dataset,
+//! Copy the given tree.
+CosineTree::CosineTree(const CosineTree& other) :
+    // Copy matrix, but only if we are the root.
+    dataset((other.parent == NULL) ? new arma::mat(*other.dataset) : NULL),
```
Nice work, this looks correct to me. :+1:

rcurtin commented on src/mlpack/core/tree/cosine_tree/cosine_tree.cpp:
```diff
@@ -150,8 +153,214 @@ CosineTree::CosineTree(const arma::mat& dataset,
+  delete right;
+
+  // Performing a deep copy of the dataset.
+  dataset = new arma::mat(*other.dataset);
```
So, either we should only allocate a new matrix here if we are the root, or always set `localDataset` to `true`. I think the first is the better approach.

sriramsk1999 commented:
Glad to hear it. Thanks for all the guidance 😄

rcurtin commented:
Thanks @sriramsk1999! :+1:
